initialization/initialization.go

```diff
@@ -141,6 +141,18 @@ func WithLogger(logger Logger) OptionFunc {
 	}
 }
 
+// withDifficultyFunc sets the difficulty function for the initializer.
+// NOTE: This is an internal option for tests and should not be used by external packages.
+func withDifficultyFunc(powDifficultyFunc func(uint64) []byte) OptionFunc {
+	return func(opts *option) error {
+		if powDifficultyFunc == nil {
+			return errors.New("difficulty function is nil")
+		}
+		opts.powDifficultyFunc = powDifficultyFunc
+		return nil
+	}
+}
```

Comment:
I'd just check against nil and panic as this is clearly a logic bug:

```suggestion
func withDifficultyFunc(powDifficultyFunc func(uint64) []byte) OptionFunc {
	if powDifficultyFunc == nil {
		panic("difficulty function is nil")
	}
	return func(opts *option) error {
		opts.powDifficultyFunc = powDifficultyFunc
		return nil
	}
}
```

Reply:
why panic instead of just erroring and falling back to the default difficulty function?

Reply:
A user might incorrectly assume that `withDifficultyFunc(nil)` they can disable looking for a nonce, analogous as to how it works with `oracle.WorkOracle`'s `WithDifficulty(d []byte)` function.

In this case `NewInitializer` should just return an error instead of panicing imo.

Reply:
It's a matter of deciding what is the valid usage of the API. If passing `nil` is clearly wrong, then a `panic` would be easier to debug / faster to catch.

Anyway, IMO, the condition should be checked at the level `withDifficultyFunc` is called, not the inner wrapper.

Reply:
I strongly disagree. `withDifficultyFunc` returns an option, that is not immediately executed. If it would also return an error a caller of that function e.g. `NewInitializer` would need to check 2 errors per parameter passed in:

current:
```golang
func NewInitializer(opts ...OptionFunc) (*Initializer, error) {
	for _, opt := range opts {
		if err := opt(options); err != nil {
			// if an option returns an error it is detected
			return nil, err
		}
	}
	// ommited for simplicity
}

// Use NewInitializer:
init, err := NewInitializer(WithOption1(5), WithOption2(nil))
```

If `withDifficultyFunc` would return an option and an error you need to do double error checking:

```golang
opt1, err := WithOption1(5)
if err != nil {
	// check
}

opt2, err := WithOption2(nil) // why should I expect this to panic?
if err != nil {
	// check
}

init, err := NewInitializer(opt1, opt2) // does error checking again, also shouldn't panic, since there is an error that can be checked
```